File: 0448-find-all-numbers-disappeared-in-an-array/0448-find-all-numbers-disappeared-in-an-array.cpp
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> result;

        // Step 1: Iterate through the array and mark visited numbers
        for (int i = 0; i < nums.size(); i++) {
            // Get the value (as an index). Use abs() because it might already be negative.
            int index = abs(nums[i]) - 1; 
            
            // Mark the value at 'index' as negative to show we've "seen" the number (index + 1)
            if (nums[index] > 0) {
                nums[index] = -nums[index];
            }
        }

        // Step 2: Iterate again to find which indices are still positive
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                // If nums[i] is positive, it means (i + 1) never appeared in the array
                result.push_back(i + 1);
            }
        }

        return result;
    }
};

File: 0448-find-all-numbers-disappeared-in-an-array/0448-find-all-numbers-disappeared-in-an-array.java
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> result = new ArrayList<>();

        for (int number : nums) {
            int value = Math.abs(number);
            int index = value - 1;
            nums[index] = -Math.abs(nums[index]); // Position becomes the presence marker.
        }

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                result.add(i + 1);
            }
        }

        return result;
    }
}

// // Can be a Solution but Here the Space complexity is O(n);
// class Solution {
//     public List<Integer> findDisappearedNumbers(int[] nums) {
//         int n = nums.length;
//         Map<Integer, Boolean> map = new HashMap<>();
//         ArrayList<Integer> result = new ArrayList<>();
//         for(int num : nums){
//             map.put(num, true);
//         }
//         for(int i =1;i<=n;i++){
//             if(map.getOrDefault(i,false)){
//                 continue;
//             }
//             result.add(i);
//         }
//         return result;
//     }
// }